src/graph/graphcontext.h

```diff
@@ -0,0 +1,78 @@
+#ifndef GRAPHCONTEXT_H
+#define GRAPHCONTEXT_H
+
+#include <assert.h>
+#include "../redismodule.h"
+#include "../index/index.h"
+#include "../stores/store.h"
+#include "graph.h"
+
+#define DEFAULT_INDEX_CAP 4
+
+typedef struct {
+  char *graph_name;                // String associated with graph
+  Graph *g;                        // Container for all matrices and entity properties
+
+  unsigned int relation_cap;       // Capacity of relation LabelStore array
+  unsigned int relation_count;     // Number of relation tyes
+  unsigned int label_cap;          // Capacity of node LabelStore array
+  unsigned int label_count;        // Number of label matrices.
+
+  LabelStore *relation_allstore;   // Schema for all relation types
+  LabelStore *node_allstore;       // Schema for all/unspecified node labels
+  LabelStore **relation_stores;    // Array of schemas for each relation type
+  LabelStore **node_stores;        // Array of schemas for each node label 
+
+  unsigned int index_cap;          // Capacity of indices array
+  unsigned int index_count;        // Number of indices
+  Index **indices;                 // Array of all indices on label-property pairs
+
+  pthread_rwlock_t _rwlock;        // Read-write lock scoped to this specific graph
+  bool writelocked;                // true if the read-write lock was acquired by a writer
+} GraphContext;
+
+/* GraphContext synchronization functions
+ * The GraphContext is initialized with a read-write lock allowing
+ * concurrent access from one writer or N readers. */
+// Acquire a lock that does not restrict access from additional reader threads
+void GraphContext_AcquireReadLock(GraphContext *gc);
+// Acquire a lock for exclusive access to this graph's data
+void GraphContext_AcquireWriteLock(GraphContext *gc);
+// Release the held lock
+void GraphContext_ReleaseLock(GraphContext *gc);
+
+/* GraphContext API */
+GraphContext* GraphContext_New(RedisModuleCtx *ctx, RedisModuleString *rs_name);
+GraphContext* GraphContext_Get(RedisModuleCtx *ctx, RedisModuleString *rs_graph_name, bool readonly);
+
+// TODO I think it might be cleaner to make the node and edge DataBlocks members of the
+// GraphContext, and use Graph as exclusively a matrix API (possibly renaming it).
+// Both would then have similar sections here.
```

Comment:
@swilly22 your thoughts here?

Reply:
I'm not sure about this, as this blurs the responsibility of the graph object,
I like to think of the GraphCTX as a container for the Graph obj, schemes and indices, I would want to fetch an entity from the GraphCTX.

Reply:
That's fair - removed this comment and the stubbed Matrix API.